### ISSUE TITLE
zapi load perf problem

### ISSUE THREAD

**Issue by mccanne** [brimdata/zed#3051]
`zapi load` is unreasonably slower than than `zed lake load`.
```
% time zed lake load -R test -use p1@main conn.log.gz 
1xzhYxGW7gqVufFNP5fW8X4LHki committed
4.345u 0.254s 0:04.31 106.4%	0+0k 0+0io 0pf+0w
% time zapi load -use p2@main conn.log.gz 
1xzhdy0WmqFePS25b0kINVde5AV committed
12.022u 20.480s 0:20.72 156.8%	0+0k 0+0io 0pf+0w
```
It should be a little slower because of the trip through the server but not this much slower!!!  We must be doing extra format conversion(s).

**Comment by mccanne** [comment]
Indeed, the problem is lake/api.Interface.Load takes a zio.Reader.  So the client decodes the input to zng records then re-encodes as ZNG, sends the ZNG to the server, the server reconstructs zng records, validates the records (after upcoming PR to fix server-path  validation), then writes the records back as ZNG to lake data objects.

I think lake/api.Interface.Load should take an io.Reader and it should just copy to the load service end point for the remote case and wrap in a zio.Reader for the local case (as the caller currently does).

That said, this is all complicated by the fact that we currently decode different input types in different files so we can't just wrap the input files in a concat reader that implement io.Reader... there is a bit of refactoring here to do to make this work.  I think it's worth it to avoid the ~3X overhead of the current design.

**Comment by philrz** [comment]
I just did a check of how this is looking in the current CLI tooling and the gap is not nearly as significant as it was when this was opened. I'm testing on my Intel-based Mac laptop. Zed is currently at commit ed3b086.

```
$ zed -version
Version: v1.8.1-55-ged3b0865
```
The intro text didn't include full detail on the test data being used, but given the `.log` suffix, the date when the issue was opened, and the mention how the client had to convert input data to ZNG, I'm guessing it was probably Zeek TSV. To make the perf delta more noticeable I therefore used the `conn.log` form the "wrccdc-year1" data set.

```
$ ls -l
...
-rw-r--r--   1 phil  staff   9576657466 Jan 22  2021 conn.log
```

Reproducing the load operations using current Zed commit ed3b086, with the input file already uncompressed the performance is 1m33s over local filesystem vs. 1m59s over localhost.

### Uncompressed Zeek TSV - Local filesystem

```
$ zed -lake lake init
lake created: lake

$ zed -lake lake create -use foo
pool created: foo 2RuCiunSBTz2xCUS0FvQ2273jn3
Switched to branch "main" on pool "foo"

$ time zed -lake lake load ~/Downloads/conn.log 
(1/1) 9.58GB/9.58GB 0B/s 100.00%
2RuCyi7g7WfPXSEfTFNNp6tueA1 committed

real	1m33.683s
user	2m43.532s
sys	0m7.669s
```

### Uncompressed Zeek TSV - Via localhost

```
$ zed create -use foo
pool created: foo 2RuDCEePOivm1lOz3Bi4ZNWraPJ
Switched to branch "main" on pool "foo"

$ time zed load ~/Downloads/conn.log 
(1/1) 9.58GB/9.58GB 0B/s 100.00%
2RuDRR0DG3gZhCDGjcQ2w5Hb5DG committed

real	1m59.555s
user	1m50.107s
sys	0m6.027s
```

---

Since @mccanne's original repro used a GZIP-compressed log file, I checked that perf as well: 2m4s over local filesystem vs. 2m31s over localhost.

### GZIP-compressed Zeek TSV - Local filesystem

```
$ zed -lake lake init
lake created: lake

$ zed -lake lake create -use foo
pool created: foo 2RuE37khRWZrsyxszop54eMZ35A
Switched to branch "main" on pool "foo"

$ time zed -lake lake load ~/Downloads/conn.log.gz 
(1/1) 2.40GB/2.40GB 0B/s 100.00%
2RuEJixPEaXph4J9qGykriqXnYm committed

real	2m4.540s
user	3m13.663s
sys	0m4.013s
```

### GZIP-compressed Zeek TSV - Via localhost

```
$ zed create -use foo
pool created: foo 2RuEWM5u3mHub5DmLQFBO82Fs7j
Switched to branch "main" on pool "foo"

$ time zed load ~/Downloads/conn.log.gz 
(1/1) 2.40GB/2.40GB 0B/s 100.00%
2RuEq35GuEeCDDT4U44XF7BWquv committed

real	2m31.618s
user	2m25.018s
sys	0m4.749s
```

---

Finally, I happened to have a version of the same log file that's GZIP-compressed ZNG where the records happen to be pre-sorted.

```
$ ls -l
...
-rw-r--r--  1 phil  staff   2943753448 Jun 20 19:01 conn-sorted.zng.gz
```

In this case there's barely any performance difference: 1m2s over local filesystem vs. 1m7s over localhost.


### GZIP-compressed ZNG - Local filesystem

```
$ zed -lake lake init
lake created: lake

$ zed -lake lake create -use foo
pool created: foo 2RuFO5RRgTXGuSg4BdIlrBZOTDD
Switched to branch "main" on pool "foo"

$ time zed -lake lake load ~/Downloads/conn-sorted.zng.gz 
(1/1) 2.94GB/2.94GB 0B/s 100.00%
2RuFWdXCQ6yAVfXJqrWk298ys6F committed

real	1m2.475s
user	2m30.608s
sys	0m4.660s
```

### GZIP-compressed ZNG - Via localhost

```
$ zed create -use foo
pool created: foo 2RuFdE3LzVENcgQoeOxEZPvBcFO
Switched to branch "main" on pool "foo"

$ time zed load ~/Downloads/conn-sorted.zng.gz 
(1/1) 2.94GB/2.94GB 0B/s 100.00%
2RuFmbS58DZmeTKBidwfV2nYbyY committed

real	1m7.760s
user	2m10.745s
sys	0m6.284s
```

---

In conclusion, we're much better off than the 3x perf delta originally observed, with the current perf drag of going through localhost & `zed serve` being around 20-30% for the Zeek TSV cases and <10% for ZNG. I'm not sure if any of @mccanne's original thoughts on addressing the remaining delta still apply to the current code base. Personally I'd be fine closing this one (i.e., if I was just seeing these perf deltas now for the first time, I'd not have thought it significant enough to open an issue) but will leave that call to the Dev team.

**Comment by philrz** [comment]
I reviewed these results with @mccanne and he expressed interest in the results of one more permutation where the ZNG is _not_ GZIP-compressed. Below are those results with the current tip of `main`, which once again show a trivial difference: 1m0s over local filesystem vs 1m5s over localhost.

```
$ ls -l
...
-rwxr-xr-x  1 phil  staff  3496836504 Jul  5 12:17 conn-sorted.zng
```

### ZNG - Local filesystem

```
$ zed -lake lake init
lake created: lake

$ zed -lake lake create -use foo
pool created: foo 2SAM26d4gakOAcuML7yP4pmTvUb
Switched to branch "main" on pool "foo"

$ time zed -lake lake load ~/Downloads/conn-sorted.zng
(1/1) 3.50GB/3.50GB 0B/s 100.00%
2SAMACAI7KnVfnhMw8l3Xr8XsTC committed

real	1m0.453s
user	1m58.916s
sys	0m5.723s
```

### ZNG - Via localhost

```
$ zed create -use foo
pool created: foo 2SAN83dWAsl0KIDotJzX3roaEz5
Switched to branch "main" on pool "foo"

$ time zed load ~/Downloads/conn-sorted.zng
(1/1) 3.50GB/3.50GB 0B/s 100.00%
2SANHmYqZZJ2FqpTLc8jpcRuwba committed

real	1m4.783s
user	1m18.730s
sys	0m5.262s
```

---

Based on this, we've agreed this issue can be closed.
